### PR TITLE
Ash's function changes

### PR DESCRIPTION
**Pull request by aofarrel**
Small function changes. Everything else is basically identical to #20 
* Filenames that contain extra periods are now supported
* Files without the extension cram or cram are now ignored instead of raising a ValueError, as metadata files are often present in bulk downloads
* Now purposely raises KeyError when there is a mismatch in the number of crams and crams and tells the user why the KeyError is being raised --> perhaps this would be better handled instead of erroring out? It can be done, [I did something like that in PT](https://github.com/aofarrel/ttt/blob/master/Paternity%20Test.ipynb), but that would require checking every input manually if this condition is triggered.

Untested due to credential issues, but it passes flake8.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [flake8]
 max-line-length=120
-ignore: E301, E302, E305, E401, F401, W503, W504, W605, E252, E402, E501, E265
+ignore: E301, E302, E305, E401, F401, W503, W504, W605, E252, E402, E501, E265, W291
```
Comment (xbrianh) [inline .flake8]: Please revert, we don't want trailing whitespace.

File: notebooks/byod.py [hunk 2]
```diff
@@ -186,7 +192,7 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
     """
 
 with callysto.Cell("python"):
-    fulldir = bucket+"/"+subdirectory
+    fulldir = bucket + "/" + subdirectory
```
Comment (xbrianh) [inline notebooks/byod.py:195]: Consider asking the user to list the path that they uploaded, which may or may not be "subdirectory".

File: notebooks/byod.py [hunk 3]
```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
+            ext = parse_cram_crai(filename)
```
Comment (xbrianh) [inline notebooks/byod.py:118]: should be:
```
sample, ext = parse_cram_crai(filename)
```

File: notebooks/byod.py [hunk 4]
```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
             if "cram" == ext:
                 crams[sample] = key
             elif "crai" == ext:
                 crais[sample] = key
             else:
                 continue
         samples = sorted(crams.keys())
+        if len(crams) != len(crais):
+            raise KeyError("Mismatch in number of cram and crai files.")
```
Comment (xbrianh) [inline notebooks/byod.py:127]: This isn't really a key error. Instead, consider:
```
assert len(crams) == len(crais), "Mismatch in number of cram and crai files."
```

Alternatively, we could find matched and unmatched crams and crais:
```
common_samples = set(crams.keys()).intersection(set(crais.keys())
unmatched_crams = set(crams.keys()).difference(set(crais.keys()))
unmatched_crais = set(crais.keys()).difference(set(crams.keys()))
if unmatched_crams:
    warnings.warn("Crams missing crais for samples {unmatched_crams}")
if unmatched_crais:
    warnings.warn("Crais missing crams for samples {unmatched_crais}")
```
and construct the table from `common_samples`.

File: notebooks/byod.py [hunk 5]
```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
+    def parse_cram_crai(filename: str):
+        sample, _ = filename.split(".", 1)                                                                    
+        if filename.endswith(".cram"): 
+            ext = "cram" 
+        elif filename.endswith(".crai") or filename.endswith(".cram.crai"):
+            ext = "crai" 
+        else:
+            # Metadata etc often takes the form of *.txt files so we'll ignore
+            # it rather than raising an exception
+            print(f"Unable to parse {filename} into sample and extension")
```
Comment (xbrianh) [inline notebooks/byod.py:110]: I think it's appropriate to silently skip non-cram-crai files.

We can add documentation somewhere indicating that other file types will be ignored.

**Review by xbrianh — CHANGES_REQUESTED** [review]
Pretty cool, love the focussed PR! A couple comments to address.

File: notebooks/byod.py [hunk 6]
```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, _ = filename.split(".", 1)
```
Comment (DailyDreaming) [inline notebooks/byod.py:102]: This just grabs the first element:

```
>>> 'a.b.c'.split(".", 1)
['a', 'b.c']
```

So a file called `1.reads.NG12345.cram.crai` would just become `1`.  This should probably be changed to only remove any of `.cram`/`.crai`/`.cram.crai`.

File: notebooks/byod.py (hunk 3 above)
Comment (DailyDreaming) [inline notebooks/byod.py:118]: `ext = parse_cram_crai(filename)`

should be:

`sample, ext = parse_cram_crai(filename)`

File: notebooks/byod.py (hunk 2 above)
Comment (aofarrel) [reply notebooks/byod.py:195]: It already does, see line 162. All that changed here from #20 is that on #27 flake8 decided that spaces are needed between operators.

File: notebooks/byod.py [hunk 7]
```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
+            ext = parse_cram_crai(filename)
             if "cram" == ext:
                 crams[sample] = key
             elif "crai" == ext:
                 crais[sample] = key
             else:
```
Comment (DailyDreaming) [inline notebooks/byod.py:123]: This is a no op and will never occur since parse_cram_crai(filename) only ever returns cram or crai.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py (hunk 3 above)
Comment (aofarrel) [reply notebooks/byod.py:118]: It has been a hot minute since I dealt with functions returning more than one value. Does this return a string of sample and a string of ext, or does it return a list containing two strings? ie does the next section still work?

Comment (xbrianh) [reply notebooks/byod.py:118]: Give it a whirl:
```
In [25]: def foo(): return 23, "blah"                                                                        

In [26]: isinstance(foo(), tuple)                                                                            
Out[26]: True

In [27]: a, b = foo()                                                                                        

In [28]: print(a, b)                                                                                         
23 blah
```

File: .flake8 (hunk 1 above)
Comment (aofarrel) [reply .flake8]: It's required to make bullet points show up correctly in markdown.

Comment (xbrianh) [reply .flake8]: In my testing, trailing whitespace is _not_ needed for bullet points. Can you provide an example?

File: notebooks/byod.py (hunk 5 above)
Comment (aofarrel) [reply notebooks/byod.py:110]: Line 46 already say this will happen so I'll take out the print.

File: notebooks/byod.py [hunk 8]
```diff
@@ -98,30 +98,39 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, ext = filename.str.rsplit(".", 1)
```
Comment (DailyDreaming) [inline notebooks/byod.py:102]: Should this return `a.cram, crai` for the crai file and `a, cram` for the cram file when the crai file is called `a.cram.crai`?

File: .flake8 (hunk 1 above)
Comment (aofarrel) [reply .flake8]: ## Assumptions
    * You are not trying to overwrite a data table that already exists  
    * Your files follow a naming convention either like this...  
    NWD119844.CRAM  
    NWD119844.CRAM.CRAI  
    ...or this:  
    NWD119844.CRAM  
    NWD119844.CRAI  

^ When I originally put this in without the trailing spaces it currently has, all of it was on the same line. Newlines don't really consistently work in markdown unless there's two of them, which looks weird when you're dealing with bulletpoints.

File: notebooks/byod.py [hunk 9]
```diff
@@ -98,30 +98,39 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
+            sample, ext = parse_cram_crai(filename)
             if "cram" == ext:
                 crams[sample] = key
             elif "crai" == ext:
                 crais[sample] = key
             else:
```
Comment (DailyDreaming) [inline notebooks/byod.py:115]: This hasn't been resolved yet.

File: notebooks/byod.py [hunk 10]
```diff
@@ -98,30 +98,39 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
                 continue
         samples = sorted(crams.keys())
-        upload_columns(table, dict(sample=samples,
-                                   cram=[crams[s] for s in samples],
-                                   crai=[crais[s] for s in samples]))
+        if len(crams) != len(crais):
+            # Finding intersections can take time, so we only do this if needed
+            common_samples = set(crams.keys()).intersection(set(crais.keys()))
```
Comment (DailyDreaming) [inline notebooks/byod.py:120]: What does:

```
            # Finding intersections can take time, so we only do this if needed
                                   crai=[crais[s] for s in samples]))
            common_samples = set(crams.keys()).intersection(set(crais.keys()))
            unmatched_crams = set(crams.keys()).difference(set(crais.keys()))
            unmatched_crais = set(crais.keys()).difference(set(crams.keys()))
```

do that the main code does not?  For example, under what situation would this occur?

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py (hunk 9 above)
Comment (aofarrel) [reply notebooks/byod.py:115]: This prevents a KeyError, doesn't it?

File: notebooks/byod.py (hunk 10 above)
Comment (aofarrel) [reply notebooks/byod.py:120]: Users bucket: 
> foo.cram
> foo.crai
> bar.cram
> sequencinginformation.txt

Resulting lists:

> crams = ["foo.cram", "bar.cram"]
> crais = ["foo.crai"]

If we don't check len(crams) != len(crais) then when we call the upload_columns function we will get a KeyError without any useful information given to the user. So we have to handle it somehow. Therefore, if there is a mismatch (and only if there is a mismatch to save computational time because in my experience this is a drag if there's 1000s of files in the bucket), we only want to upload the intersection -- ie, where there are crais for each cram. Technically you don't *need* to do avoid uploading blank rows as the resulting TSV will still be valid, but it could break some workflows if a .crai file is expected.

File: notebooks/byod.py (hunk 9 above)
Comment (DailyDreaming) [reply notebooks/byod.py:115]: The final `else` is unnecessary.  Run this with and without the commented part:

```
crams = dict()
crais = dict()

files = ['a.cram', 'a.crai', 'b.baloney', 'c.crai', 'c.cram']


def release_the_files(files):
    for file in files:
        if file.endswith('cram'):
            crams[file] = 'a'
        elif file.endswith('crai'):
            crais[file] = 'c'
        # else:
        #     continue
    print('I return!')
    return crams, crais


crams, crais = release_the_files(files)

print(crams, crais)
```

Comment (DailyDreaming) [reply notebooks/byod.py:115]: Note, the change could also be done in a separate PR.

File: notebooks/byod.py (hunk 8 above)
Comment (aofarrel) [reply notebooks/byod.py:102]: Yes, that's intended, although I just realized that will cause issues during upload_tables(). I'll mull this over and try to think of a good fix that can handle both cases.

File: notebooks/byod.py (hunk 9 above)
Comment (aofarrel) [reply notebooks/byod.py:115]: Oh, I see. I always add a terminal else when I also use elif out of habit. I think my Java professor drilled it into me even though it's not required. I can remove it though.

File: notebooks/byod.py (hunk 10 above)
Comment (DailyDreaming) [reply notebooks/byod.py:121]: Since you're already doing the matching earlier, would it be cheaper to keep track of the unmatched elements as you go?  I see this potentially taking a long time reiterating all of the crams and all of the crais another three times each.

So instead of something like this:

```
crams = {}
crais = {}
for i in large_list:
    if cram:
        crams.add(i)
    elif crai:
        crais.add(i)

if not_exactly_equal:
    loop over all crai
    loop over all cram

    and again...

    loop over all crai
    loop over all cram

    and again...

    loop over all crai
    loop over all cram
```


Writing a function that only loops over things once (or twice), the first time?

Comment (aofarrel) [reply notebooks/byod.py:121]: The matching isn't being done earlier, though?

Comment (aofarrel) [reply notebooks/byod.py:121]: Oh, wait, I see what you mean. The matching within that if block. Yep, I can fix that!

Comment (DailyDreaming) [reply notebooks/byod.py:121]: For example, you could use something like:

```
from collections import defaultdict
matched_files = defaultdict(dict())

for filename in large_list:
    basename, ext = split_up_filename(filename)
    if ext in ['cram', 'crai']:
        matched_files[basename][ext] = filename

for matched_file in matched_files:
    if matched_file.get('cram') and matched_file.get('crai'):
        print('These two are paired.')
    elif matched_file.get('cram') and not matched_file.get('crai'):
        print(f'{cram} is missing an index.')
    elif not matched_file.get('cram') and matched_file.get('crai'):
        print(f'{crai} is missing a cram file.')
```

Comment (DailyDreaming) [reply notebooks/byod.py:121]: @aofarrel This was closed but my point was that `.intersection` and `.difference` should never need to be called.

File: notebooks/byod.py [hunk 11]
```diff
@@ -98,30 +98,41 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, ext = filename.rsplit(".", 1)
+        if sample.endswith("cram"):
```
Comment (DailyDreaming) [inline notebooks/byod.py:103]: So... what if someone has a file named: `oh.my_silly_cram.cram`?

This function should probably, explicitly, check and remove what it's expecting.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py (hunk 10 above)
Comment (aofarrel) [reply notebooks/byod.py:121]: My b, I hit resolve just as you commented.

File: notebooks/byod.py (hunk 11 above)
Comment (DailyDreaming) [reply notebooks/byod.py:103]: This was marked resolved, but was I see no commit afterwards?

Comment (aofarrel) [reply notebooks/byod.py:103]: Because the fix was just a one character change. It'll be included in the commit that includes defaultdict.

File: .flake8 (hunk 1 above)
Comment (aofarrel) [reply .flake8]: <br> is supposed on Jupyter, swapping to that instead

File: notebooks/byod.py [hunk 12]
```diff
@@ -98,30 +99,51 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
```
Comment (DailyDreaming) [inline notebooks/byod.py:102]: Please remove `rsplit()` from this function as it produces edge case errors and explicitly only remove `.cram` or `.crai` or `.cram.crai`.

File: notebooks/byod.py [hunk 13]
```diff
@@ -98,30 +99,51 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
-                                   crai=[crais[s] for s in samples]))
+            sample, ext = parse_cram_crai(filename)
+            if ext in ['cram', 'crai']:
+                matched_files[sample][ext] = filename
+                if ext == 'cram':
+                    numCrams += 1
+                else:
+                    numCrais += 1
+
+        if numCrais != numCrams:
```
Comment (DailyDreaming) [inline notebooks/byod.py:123]: What if the list has 10 unmatched crams, and 10 unmatched crais?

File: notebooks/byod.py [hunk 14]
```diff
@@ -98,30 +99,51 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
+            # so this only runs if there's more crams than crais or vice versa
+            for matched_file in matched_files:
+                if matched_file.get('cram') and not matched_file.get('crai'):
+                    warnings.warn(f'{cram} is missing an index.')
+                    del matched_files[sample]
+                elif not matched_file.get('cram') and matched_file.get('crai'):
+                    warnings.warn(f'{crai} is missing a cram file.')
+                    del matched_files[sample]
+        
+        crams = []
```
Comment (DailyDreaming) [inline notebooks/byod.py:134]: `crams = []` is a list.  Why do you call `crams[ext]` below?

```
        crams = []
        crais = []
        for sample in matched_files:
            for ext in ('cram', 'crai'):
                if ext == 'cram':
                    crams.append(matched_files[sample][ext])
                else:
                    crais.append(matched_files[sample][ext])

        # Upload TSV
        upload_columns(table, dict(sample=[samples[s] for s in matched_files],
                cram=[crams[ext] for ext in ('cram', 'crai')],
                crai=[crais[ext] for ext in ('cram', 'crai')]))
```

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py [hunk 15]
```diff
@@ -98,40 +100,65 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
+                    allCrams.append(sample)
+                else:
+                    allCrais.append(sample)
+
+        if allCrams != allCrais:
+            # We don't want to iterate an additional time if we don't need to,
+            # so this only runs if there's more crams than crais or vice versa
+            for sample in list(matched_files):
+                try:
+                    if matched_files[sample]['cram'] and not matched_files[sample]['crai']:
```
Comment (DailyDreaming) [inline notebooks/byod.py:129]: Will:

```
                    if matched_files[sample]['cram'] and not matched_files[sample]['crai']:
                        if not matched_files[sample]['cram'] and matched_files[sample]['crai']:
```

ever occur?

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py (hunk 15 above)
Comment (aofarrel) [reply notebooks/byod.py:129]: Either one of those lines can throw a KeyError, yes. It'll be the first line if there is no crai and the second line if there is no cram. They're nested because if they were not, this happened:
fizz is missing a crai --> keyerror, deletes fizz --> fizz is missing a cram --> keyerror again
Instead of...
fizz is missing a crai --> keyerror, deletes fizz --> moves on

Oddly enough I didn't get KeyErrors earlier on; you'll see older commits handled this better because without KeyErrors I can account for either case. I think something is wrong with my defaultdictionary setup; I had to go from defaultdict(dict()) to defaultdict() at some point.

Comment (DailyDreaming) [reply notebooks/byod.py:129]: Have you heard of `dict().get(item)`?  As it is, this code is an impossible conditional.

Comment (aofarrel) [reply notebooks/byod.py]: Found a solution. It's not exactly elegant, though. --> https://github.com/DataBiosphere/bdcat_notebooks/pull/27/commits/93f0b4a780db92fcda3e982bd4943c78f2e5a0da

File: .gitignore [hunk 16]
```diff
@@ -6,3 +6,5 @@ __pycache__
 .idea/
 .coverage*
 .mypy_cache
+notebooks/.DS_Store
```
Comment (DailyDreaming) [inline .gitignore:9]: This should be a separate PR, as Brian mentioned.

File: notebooks/byod.py [hunk 17]
```diff
@@ -1,17 +1,18 @@
-# publish to: "terra-notebook-utils-tests" "test-byod"
+#publish to: "TEMP-byod-edits" "test-byod-local"
 import os
 import callysto
 
 # Mock the notebook environment
-os.environ['WORKSPACE_NAME'] = "terra-notebook-utils-tests"
-os.environ['WORKSPACE_BUCKET'] = "gs://fc-9169fcd1-92ce-4d60-9d2d-d19fd326ff10"
-os.environ['GOOGLE_PROJECT'] = "firecloud-cgl"
+os.environ['WORKSPACE_NAME'] = "TEMP-byod-edits"
```
Comment (DailyDreaming) [inline notebooks/byod.py:6]: Why are these being changed?

File: notebooks/byod.py [hunk 18]
```diff
@@ -72,134 +98,136 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
-                                   crai=[crais[s] for s in samples]))
+            sample, ext = parse_cram_crai(filename)
+            if ext in ['cram', 'crai']:
+                matched_files[sample][ext] = filename
+                if ext == 'cram':
+                    allCrams.append(sample)
+                else:
+                    allCrais.append(sample)
+
+        if allCrams != allCrais:
```
Comment (DailyDreaming) [inline notebooks/byod.py:124]: This still seems like a functional issue where if you have the same number of unmatched crams and crais, this will error.

File: notebooks/byod.py [hunk 19]
```diff
@@ -72,134 +98,136 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
...
+            # We don't want to iterate an additional time if we don't need to,
+            # so this only runs if there's more crams than crais or vice versa
+            for sample in list(matched_files):
+                try:
+                    if not matched_files[sample]['cram']:
```
Comment (DailyDreaming) [inline notebooks/byod.py:129]: Have you heard of matched_files[sample].get('cram')?  I think it might be easier and cleaner than the `try` / `except` logic throughout this.

**Review by DailyDreaming — CHANGES_REQUESTED** [review]

File: notebooks/byod.py (hunk 18 above)
Comment (aofarrel) [reply notebooks/byod.py:122]: I don't think this is an issue unless I don't understand what you mean. My test case actually is exactly that scenario (one sample has a missing cram and another has a missing crai) and it always catches it. `allCrams != allCrais` isn't equivalent to what I was doing earlier, which was basically `len(allCrams) != len(allCrais)`

File: notebooks/byod.py (hunk 17 above)
Comment (aofarrel) [reply notebooks/byod.py:6]: 1. To push to my own workspace rather than clutter up Brians
2. To avoid Brian and I pushing notebooks on top of each other and not being able to keep track of whose is whose
3. To debug my persistent credential issues in a fresh workspace where I have owner access
4. To debug ' ' ' with a billing project that I'm allowed use directly, unlike firecloud-cgl

File: .gitignore (hunk 16 above)
Comment (aofarrel) [reply .gitignore]: Not quite sure I understand why? .DS_Store is a binary file used on Mac to keep track of the Finder UI. It changes when you so much as blink and regenerates itself when deleted. If not in .gitignore, it'll keep regenerating itself and try to add itself to every commit I make, so this was for my own sanity and to keep the remote repo free of unneeded files.

File: notebooks/byod.py (hunk 19 above)
Comment (aofarrel) [reply notebooks/byod.py:127]: If I understand what you're suggesting to be...
```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            if not matched_files[sample].get('cram'):
                continue
            print(f'{sample} is missing a cram file and will be removed')
            del matched_files[sample]
            break
            if not matched_files[sample].get('crai'):
                continue
            print(f'{sample} is missing a crai file and will be removed')
            del matched_files[sample]
```
...then that won't work, it'll appear to pass but then throw a KeyError when creating the table to be uploaded. Neither does this, although I'm not quite sure why, seeing as it keeps the try/except blocks.

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            try:
                if not matched_files[sample].get('cram'):
                    continue
            except KeyError:
                print(f'{sample} is missing a cram file and will be removed')
                del matched_files[sample]
                break
            try:
                if not matched_files[sample].get('crai'):
                    continue
            except KeyError:
                print(f'{sample} is missing a crai file and will be removed')
                del matched_files[sample]
```

Comment (DailyDreaming) [reply notebooks/byod.py:136]: It would actually be something like this:

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            if not (matched_files[sample].get('cram') or matched_files[sample].get('crai')):
                del matched_files[sample]
```

**Review by DailyDreaming — COMMENTED** [review]
Added one comment on the try/except blocks.  Once that's addressed, I think that this is good.

File: notebooks/byod.py (hunk 19 above)
Comment (aofarrel) [reply notebooks/byod.py:136]: That throws a KeyError later down the line.

File: notebooks/byod.py (hunk 18 above)
Comment (DailyDreaming) [reply notebooks/byod.py:131]: The comparison actually does an iteration of all elements in order to compare the two lists, but that's more of a nit.  This is fine.

File: .gitignore (hunk 16 above)
Comment (DailyDreaming) [reply .gitignore]: You have to `git add /specific/file` when it's not in the ignore file.  Don't always use `git add .`.

File: notebooks/byod.py (hunk 19 above)
Comment (DailyDreaming) [reply notebooks/byod.py:136]: This:

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            try:
                if not matched_files[sample]['cram']:
                    continue
            except KeyError:
                print(f'{sample} is missing a cram file and will be removed')
                del matched_files[sample]
                break
            try:
                if not matched_files[sample]['crai']:
                    continue
            except KeyError:
                print(f'{sample} is missing a crai file and will be removed')
                del matched_files[sample]

        crams = []
        crais = []
        samples = []
        for sample in matched_files:
            samples.append(sample)
            for ext in ('cram', 'crai'):
                if ext == 'cram':
                    crams.append(f"{bucket}/{subdirectory}/{matched_files[sample][ext]}")
                else:
                    crais.append(f"{bucket}/{subdirectory}/{matched_files[sample][ext]}")
```

Seems to be the same as this:

```
for sample in list(matched_files):
    if matched_files[sample].get('cram') and matched_files[sample].get('crai'):
        crams.append(f"{bucket}/{subdirectory}/{matched_files[sample]['cram']}")
        crais.append(f"{bucket}/{subdirectory}/{matched_files[sample]['crai']}")
    else:
        print(f'{sample} is missing a file and will be removed:\n{json.dumps(matched_files[sample])}')
```

**Review by DailyDreaming — APPROVED** [review]
Looks great!  Thanks for addressing all of the comments!

File: notebooks/byod.py [hunk 20]
```diff
@@ -105,30 +104,52 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        if filename.endswith(".cram.crai"):
+            # double extension in the crai file, ie "foo.cram.crai"
+            sample = filename[:-10]
+        else:
+            sample = filename[:-5]
+        ext = filename[-4:]
```
Comment (xbrianh) [inline notebooks/byod.py:113]: Assuming extension lengths looks brittle, but this can be handled in a future work.

**Review by xbrianh — APPROVED** [review]
